core/dbt/parser/read_files.py

```diff
@@ -153,104 +174,244 @@ def generate_dbt_ignore_spec(project_root):
     return ignore_spec
 
 
-# This needs to read files for multiple projects, so the 'files'
-# dictionary needs to be passed in. What determines the order of
-# the various projects? Is the root project always last? Do the
-# non-root projects need to be done separately in order?
-def read_files(project, files, parser_files, saved_files):
-    dbt_ignore_spec = generate_dbt_ignore_spec(project.project_root)
-    project_files = {}
-
-    project_files["MacroParser"] = read_files_for_parser(
-        project,
-        files,
-        project.macro_paths,
-        [".sql"],
-        ParseFileType.Macro,
-        saved_files,
-        dbt_ignore_spec,
-    )
+@dataclass
+class ReadFilesFromFileSystem:
```

Comment:
Just generally, more typing would have helped me understand more of this file more quickly. That said, I think the best way to capture shared behavior between classes like ReadFilesFromFileSystem and ReadFilesFromDiff is by defining a Protocol. But Protocols were introduced in Python 3.8, so we have a few more months without them.

Reply:
We do use protocols in a few places already, so I think there are compatibility things for before 3.8.